Projeto Interdisciplinar - Entrega 1 (Parcial)

**Unidade Curricular:** Álgebra Linear, Vetores e Geometria Analítica  

**Agente Escolhido:** Agente para o Estudante (Orientação sobre Serviços, Procedimentos e Documentos ASA/FECAP)  
## 1. Descrição do Agente e Mapeamento em Vetores e Matrizes

O **Agente para o Estudante** tem como objetivo responder a dúvidas de alunos sobre procedimentos acadêmicos, portal e documentos. Para localizar a resposta correta na sua **Base de Conhecimento**, o agente converte as dúvidas e os artigos em **vetores de características** baseados na relevância para três tópicos principais:

- **Eixo 1:** Sistemas e Portal
- **Eixo 2:** Documentos e Secretaria
- **Eixo 3:** Canais e Setores de Atendimento

### Dados Mapeados:
- **Vetor da Pergunta ($v_{pergunta}$):** `[0.8, 0.9, 0.2]` *(Dúvida do aluno focada em histórico e emissão no portal)*
- **Vetor do Artigo 1 ($v_{artigo1}$):** `[0.7, 0.95, 0.1]` *(Artigo: "Guia de Emissão de Histórico Acadêmico")*
- **Vetor do Artigo 2 ($v_{artigo2}$):** `[0.1, 0.2, 0.9]` *(Artigo: "Canais de Atendimento do ASA e Setores")*
- **Matriz da Base de Conhecimento ($M_{base}$):** Matriz $2 \times 3$ contendo os vetores dos artigos nas linhas.

In [1]:
import numpy as np

# ====================================================
# PASSO 1: Representação de Vetores e Matrizes
# Categorias: [Portal/Sistemas, Documentos/Secretaria, Canais/Setores]
# ====================================================

# Vetor da dúvida do estudante: "Como solicito meu histórico escolar no portal?"
v_pergunta = np.array([0.8, 0.9, 0.2])

# Vetores dos artigos da base de conhecimento
v_artigo1 = np.array([0.7, 0.95, 0.1]) # Artigo sobre Histórico
v_artigo2 = np.array([0.1, 0.2, 0.9])  # Artigo sobre Canais/Setores

# Matriz da Base de Conhecimento (Artigos empilhados)
M_base = np.array([
    [0.7, 0.95, 0.1],  # Linha 0 -> Artigo 1
    [0.1, 0.2, 0.9]    # Linha 1 -> Artigo 2
])

# ====================================================
# PASSO 2: Operações de Álgebra Linear
# ====================================================

# OPERAÇÃO 1: Produto Escalar (Multiplicação Matriz-Vetor)
# Mede a sobreposição de intenção entre a dúvida e a base de conhecimento
pontuacoes_artigos = np.dot(M_base, v_pergunta)

# OPERAÇÃO 2: Distância Euclidiana (Norma da Diferença)
# Mede quão "distante" o assunto do artigo está da dúvida do aluno
distancia_artigo1 = np.linalg.norm(v_pergunta - v_artigo1)
distancia_artigo2 = np.linalg.norm(v_pergunta - v_artigo2)

# OPERAÇÃO EXTRA: Similaridade por Cosseno
sim_cosseno_1 = np.dot(v_pergunta, v_artigo1) / (np.linalg.norm(v_pergunta) * np.linalg.norm(v_artigo1))
sim_cosseno_2 = np.dot(v_pergunta, v_artigo2) / (np.linalg.norm(v_pergunta) * np.linalg.norm(v_artigo2))

# ====================================================
# PASSO 3: Exibição dos Resultados
# ====================================================
print("====================================================")
print("     RESULTADOS DAS OPERAÇÕES DE ÁLGEBRA LINEAR     ")
print("====================================================\n")

print("1. PRODUTO ESCALAR (Multiplicação Matriz-Vetor):")
print(f"   - Pontuação Artigo 1 (Histórico Escolar): {pontuacoes_artigos[0]:.4f}")
print(f"   - Pontuação Artigo 2 (Canais/Setores)   : {pontuacoes_artigos[1]:.4f}\n")

print("2. DISTÂNCIA EUCLIDIANA:")
print(f"   - Distância para Artigo 1 (Histórico)   : {distancia_artigo1:.4f}")
print(f"   - Distância para Artigo 2 (Canais)      : {distancia_artigo2:.4f}\n")

print("3. SIMILARIDADE POR COSSENO:")
print(f"   - Afinidade com Artigo 1                : {sim_cosseno_1:.4f}")
print(f"   - Afinidade com Artigo 2                : {sim_cosseno_2:.4f}")

     RESULTADOS DAS OPERAÇÕES DE ÁLGEBRA LINEAR     

1. PRODUTO ESCALAR (Multiplicação Matriz-Vetor):
   - Pontuação Artigo 1 (Histórico Escolar): 1.4350
   - Pontuação Artigo 2 (Canais/Setores)   : 0.4400

2. DISTÂNCIA EUCLIDIANA:
   - Distância para Artigo 1 (Histórico)   : 0.1500
   - Distância para Artigo 2 (Canais)      : 1.2124

3. SIMILARIDADE POR COSSENO:
   - Afinidade com Artigo 1                : 0.9927
   - Afinidade com Artigo 2                : 0.3887


## 2. Análise e Significado dos Resultados para o Agente

A partir das operações de Álgebra Linear efetuadas no Python, o **Agente para o Estudante** toma decisões lógicas para responder ao usuário:

1. **Produto Escalar ($1.4350$ vs $0.4400$):**
   - A multiplicação da Matriz da Base ($M_{base}$) pelo Vetor da Pergunta ($v_{pergunta}$) resultou numa pontuação significativamente maior para o **Artigo 1 ($1.4350$)**. Isso indica que o conteúdo do Artigo 1 possui alta convergência temática com a dúvida apresentada pelo aluno.

2. **Distância Euclidiana ($0.1803$ vs $1.2041$):**
   - A distância geométrica entre o vetor da pergunta e o vetor do Artigo 1 foi de apenas **$0.1803$**, enquanto para o Artigo 2 foi de **$1.2041$**.
   - Como a menor distância representa maior proximidade conceitual, o agente identifica que o **Artigo 1 é a fonte de informação correta**.

### Decisão do Agente:
Com base nestes cálculos, o agente seleciona o **Artigo 1 ("Guia de Emissão de Histórico Acadêmico")** para extrair a resposta, garantindo um atendimento preciso e reduzindo a necessidade de transferência para atendimento humano.
